_posts/2021-03-09-bezier-fitting.md

```diff
@@ -0,0 +1,196 @@
+
+Cubic beziers are by far the most common curve representation, used both for design and rendering. A fundamental problem is *curve fitting,* or determining the cubic bezier that's closest to some source curve. Applications include simplifying existing paths, efficiently representing the [parallel curve], and rendering other spline representations such as Euler spiral or [hyperbezier]. A specific feature where curve fitting will be used in [Runebender] is deleting a smooth on-curve point. The intent is to merge two beziers into one, which is another way of saying to find a bezier which most closely approximates the curve formed of the two existing bezier segments.
+
+In spite of the importance of the curve-fitting problem and literature going back more than 30 years, there has been until now no fully satisfactory solution. Existing approaches either miss the optimum solution, are very slow (unsuitable for interactive use), or both.
+
+The main reason the problem is so hard is a basic but little-known fact about cubic beziers: C-shaped beziers come in triplets of very similar shape but quite different parameters. As a result, in general when fitting some arbitrary source curve, there are three local minima. Approaches based on iterative approximation (in addition to being slow) are likely to find just one, potentially missing a closer fit of one of the others.
+
+## How close are two curves?
+
+Before getting into the solution, we'll need to state the problem more carefully. The goal is a cubic bezier that minimizes some error metric with respect to the source curve. But how do we measure the distance between two curves?
+
+One meaningful metric is the [Fréchet distance]. It can be explained with a little story. An aircraft engineer goes on a walk with their dog Pierre, and the path followed by each can be described by a curve. What length leash is required so that the dog is always within that distance of the engineer? That is the Fréchet distance.
+
+Also note that the Fréchet distance is closely related to the Hausdorff distance, but the two can differ when the path loops back or crosses itself. For stroked curves, the Hausdorff distance is more relevant, but for filled paths such as font outlines, Fréchet preserves winding number in a way that Hausdorff doesn't guarantee.
+
+For solving optimization problems, the Fréchet distance is not ideal, as it effectively measures the distance of a single point, the maximum error. One can informally describe an [L2] error metric as similar, except that instead of describing the maximum length of the leash, it describes an *effort* of handling the dog which is proportional to the square of the distance of the dog from the engineer. Minimizing an L2 error means minimizing that total effort. This error metric takes into account the entire path, and thus is smoother in response to small changes of the parameters.
```

Comment:
```suggestion
For solving optimization problems, the Fréchet distance is not ideal, as it effectively measures the distance of a single point, the maximum error. One can informally describe an [L2 error metric] as similar, except that instead of describing the maximum length of the leash, it describes an *effort* of handling the dog which is proportional to the square of the distance of the dog from the engineer. Minimizing an L2 error means minimizing that total effort. This error metric takes into account the entire path, and thus is smoother in response to small changes of the parameters.
```

I try to avoid super short link stubs 🤷